### DISCUSSION TITLE
VF2++ preliminary testing

### DISCUSSION THREAD

**Discussion by kpetridis24** [networkx/networkx#5579] [General]
I have finished reading all the papers related to VF2 and VF2++ and i am currently implementing helper functions, regarding the node matching order, as determined by VF2++. After i make sure that they work properly, i am going to apply them to random graphs, using the already existing implementation of VF2 and compare the performance of the old implementation, but using the new, candidate node pair ordering. After this, i will proceed with the implementation of the new cutting rules that VF2++ proposes, and try to integrate them. This process should give a better understanding of how the algorithm works, as well as problems that could possibly emerge. I will provide any updates here and i will ask any questions so we can discuss them.  

I am adding the necessary links here for
- [VF2](https://ieeexplore.ieee.org/document/1323804)
- [VF2++](https://www.researchgate.net/publication/324096256_VF2-An_improved_subgraph_isomorphism_algorithm)

**Comment by kpetridis24** [comment]
My progress on the algorithm can be [found here](https://github.com/kpetridis24/networkx) on the **VF2++** branch, under **algrithms/isomorphism/**. The file is named isomorphvf2pp.py.

**Comment by kpetridis24** [comment]
So, regarding the candidate node pairs, the VF2++ paper suggests that we do not need to maintain P(s) (the list of candidates) but rather compute the candidate node v from G2, given a node u from G1, every time. The logic is the following: Given a node uεG1, the candidate v from G2 should be one of the neighbors of m(u'), where m is the mapping and u' is the covered neighbor of u. This makes total sense, however my question refers to the first iteration of the algorithm, when no mapping has been established and m is empty. We have found the node ordering for G1 (via the VF2++ preprocessing) and now, given a node u from the ordering, we need to get the candidate from G2 and examine the pair. u however has no covered neighbors yet so we cannot compute the neighbors of m(u'). How is this going to work if the mapping is empty? Should we take a random node from G2 to start with? @MridulS @dschult @boothby

**Comment by dschult** [comment]
My understanding is that the first constraint on nodes chosen from G2 is the degree (and the label if we are using labels). So, when you look at nodes from G2 you really are only looking at the nodes with a certain degree and label. Then if you don't have neighbor-of-u information, you pick from among the set within G2 that matches these other constraints.  Also, it may not need to be a random selection -- I'm not sure. Arbitrary selection may be as good as random. VF2++ is better about ordering the nodes before-hand. Isn't there a node ordering specified for G2? Or is the node ordering only specified for G1. It sounds like you are saying that it is only for G1.

**Reply by kpetridis24** [reply]
I read the paper a lot of times but i cannot find a section were this is explicitely specified. From my understanding, G2 should be ordered as well based on degree (and labels if they are used), and we should initiate the algorithm with two arrays containing the ordered nodes from G1 and G2 respectively. This should be the order favoring the nodes, that have a greater possibility to establish a successful mapping. 

So what you are suggesting is that, if no neighbor-of-u is present in the mapping m of the current state s, we should just take as candidate, a node from the ordering of G2 (which is not arbitrary). This divides the candidate selection process in two separate cases. The first is just like VF2 if i am not mistaken (selection of a node from T2_inout), but now from the T_ordered. The second one is also not explained in detail. The way i am thinking about it is, given a node u from G1, and m (mapping) we do
```
for n in neighborhood(u):
  if n in m:
    for candidate in neighborhood(m[n]):
      if candidate not in m:         
        examine(u, candidate)
          if mapping is extended:
            break (take next node from G1 and run for the updated state)
```
Am i missing something?

**Reply by dschult** [reply]
I have to admit, the correct logic here is not on the tip of my thoughts. You are saying there isn't a section of the paper that describes this. Can you give me a hint where in the paper to look?  

My attempt to work through that code makes me wonder whether `candidate` can possibly be in `m`.  The keys of `m` are nodes in G1 right? and `candidate` is a node from G2?

**Reply by kpetridis24** [reply]
The candidate node selection (for VF2++) is described in section 5.3 (page 9) of [this paper](https://www.researchgate.net/publication/324096256_VF2-An_improved_subgraph_isomorphism_algorithm) and for VF2 in section 2.1 of [this one](https://ieeexplore.ieee.org/document/1323804). 

I don't think that candidate can be in m, because that means that it is an already covered node, which was successfully matched in a higher level of the mapping tree. The candidates are two uncovered nodes that do not exist in m, and are being examined so that if a matching is established, they are added to m, and we search deeper into that branch. 

Yes, the keys of m are nodes of G1. This means that, given a node u from G1 and its neighbor n, m[n] (given that n is covered) gives the node v from G2 which is the node that n was matched to, in previous candidate examination (m[n] = v means n from G1 was matched to v from G2).

The main issue is that, contrarily to the VF2 paper, the one discussing VF2++ does not present the exact algorithm, so we have the pieces of the puzzle and we have to put them together to form the final algorithm. The first part concerning the node ordering is clear to me and i have already tested it. So the next step is to figure out how is the ordering going to be used exactly, in every step of the algorithm. After that, we just take a pair of candidate nodes, apply the cutting rules and extend the mapping in case of feasibility, or examine another candidate pair from another branch of the tree.

**Reply by dschult** [reply]
OK... Thanks for that. 
If the candidate cannot be in m then that part of the overview code (line 4 and onward) you provide 2 comments ago isn't relevant.

I agree that section 5.3 of the paper doesn't provide direct description of the order in which to choose candidates v from G2 to matching with node u from G1.  But here's what I get from the second and third sentence (which just restates the second): each already covered neighbor of u in G1 has to be mapped to an already covered neighbor of v in G2. Turning that around, given u in G1, we want to consider all up-to-now uncovered nodes v in G2 that match the degree/label of u in G1 **and** v must have all its covered neighbors have inverse-mapping u' that are neighbors of u.  If there aren't any covered neighbors of v then this criteria won't rule out v as a candidate. 

This does seem to narrow down the candidates pretty quickly once the mapping gets started. For the next node in the ordering of G1, you look at covered neighbors u' of u, and take the intersection of the neighborhoods in G2 of the v'=m(u').  And then take the intersection with the nodes in G2 that match the degree/label of u.  Furthermore, the VF2++ paper states that you pick **arbitrarily** from that collection of nodes. 

That's my interpretation of what that paragraph is saying...  Note: we don't need to order the nodes in G2. And Note: when the node u has no neighbors covered (e.g. when starting) we only have the nodes in G2 that match the degree/label of u in G1.

Does this make any sense?

**Reply by kpetridis24** [reply]
Thanks a lot for the answer. I will read it closely tomorrow and give more feedback/questions if necessary!

**Reply by kpetridis24** [reply]
Ok so first of all, i still believe that line 4 and onward (of the overview code) is relevant, because, for a node u of G1 we consider the neighbors of m[n] where n is a covered neighbor of u. However, we have to exclude the neighbors of m[n] which are already covered (because we are looking for candidates, i.e. nodes not present in the mapping), so line 4 discards the nodes that are already mapped. 

I agree with the second paragraph and this is exactly what i had in mind. What you are describing is basically what i try to accomplish with the overview code, but in reverse. Given the node u from G1 i look for nodes v (which are the neighbors of m[u'], u' being a covered neighbor of u), such that they are not currently in the mapping. 

Yes it makes sense. I am a bit confused however about combining all those features together, so i will take some time to read all the details that you provided again and make sure everything is clear. After that, i will try to write a pseudocode for the process and ask for feedback again. 

**Thanks a lot!**

**Reply by kpetridis24** [reply]
Regarding the third paragraph of your last response, what i understand is this:

Let u be a node from the ordering of G1. Let's assume that u has two neighbors (for simplicity), u1 and u2. These are covered, so they map to v1=m[u1] and v2=m[u2] of G2. v1 and v2 have one neighborhood of nodes each, let's name them nv1 and nv2. You suggest that i take the intersection of these neighborhoods, so nv = intersect(nv1, nv2) and then from the result of the intersection (nv), keep only those nodes that have the same degree/label as u, and call them candidates. 

Is that right?
Also, how did you come up with the intersection thing? I don't see this stated anywhere in the paper, so i just thought that we should examine every (u,v) pair (v is the neighbor of m[u']) for every u' until we find a successful matching. The intersection keeping the nodes with the same degree/label makes sense, however i cannot see why would we apply the first intersection, i.e. keep the nodes that are common neighbors of m[u'] for all u' of u. Could you please explain it a bit more?

**Reply by dschult** [reply]
I'll try to explain a bit more, but keep in mind that I am not an expert here -- just writing what I think of when I read their paragraph. So you may be able to come up with something that matches what they say better than I do.

Using your notation, v1 and v2 are the nodes in G2 mapped to by neighbors u1 and u2 of the node u we want to map.   To be the isomorphism, the node u should map to a neighbor of v1. (neighbors map to neighbors).  But it should also be a neighbor of v2. So we have to pick it from the nodes that are neighbors of both v1 and v2. That's where I came up with the intersection.  In the Juttner paper section 5.3 the sentence is: "Each covered neighbor of u has to be mapped to  neighbor of v, ..., all of the admissible candidates for u are among the neighbors of m(u')." Turning that sentence around, v must come from the nodes which are neighbors to **all** the nodes m(u') where u' are the already mapped neighbors of u. 

Hope this helps.

**Reply by kpetridis24** [reply]
**Thank you** for the detailed explanation, i think this makes more sense now and i have to agree with this as well. I am now confident that we indeed need the first intersection. 

I think i have made a significant part of the algorithm clear and i present the high-level algorithm in a generalized form (as i am thinking it at the moment). Feel free to provide feedback so we can gradually build the algorithm, correcting and optimizing the way it should work.

```
Input:  G1, G2
1.  Check if |V1|==|V2| 
2.  Check if degree_sequence(G1) == degree_sequence(G2)
3.  order1 = orderVF2pp(G1)                    // Order the nodes of G1 as [ u(0), u(1), ..., u(|V1|-1) ]
4.  M = { }                                    // M[i] = v if u(i) matches with v, u(i) from ord1
5.  for i, u in enumerate(ord1):
6.    v in find_candidate(u)
7.    if feasibility(u, v, s):
8.      s <- s U (u, v)                        // s: M[i] = v
9.      take next u from ord1
```

```
1. def find_candidate(u):
2.   C = [ ] 
3.   if neighborhood(u) is empty:               // if u has no neighbors, just take G2 nodes with same degree as u
4.     C = [ nodes from G2 that are not covered yet and have the same degree as u ]
5.   else:
6.     C = compute intersection of neighborhoods in G2, of every v, such that m[u'] = v, u' being a neighbor of u in G1.
8.     C = [ nodes from C that have the same degree as u ]
9.   return C
```

This should describe a significant portion of how the algorithm should work, given that there are no labels (for simplicity). If you believe that something here should be corrected, feel free to point out!

**Reply by dschult** [reply]
This looks good -- though I have one comment and a couple of nitpicky things:
- you'll have to use some sort of queue (or recursive functions but they are slow and take memory) to do step 9. And that will require some clean-up code after you are done with step 9 to remove `M[u]=v` from the mapping and try another of the candidates.  So, the whole 5-9 loop will probably be inside a big while loop putting things on a queue and taking them off until the queue is empty.  What I guess I mean is that step 9 looks like it is part of the for loop in step 5 -- but there is more involved there. We don't have to figure all the out now. But we will need some sort of ability to try the next candidate and when we finally find that it doesn't work, take it back out again.
- you shouldn't need `i`. You can let the mapping be `M[u] = v`.  Then all indexes will be nodes (we don't need to know the order for the v).
- You introduce `s` in 7. and 8. but I think that can just be `M[u] = v` (and updating any inverse mapping if you need that for feasibility) Will you need to know the inverse mapping?

**Reply by kpetridis24** [reply]
Regarding the second bullet point, i just used i to be consistent with the paper notation. Of cource we could use every node as a key to M and not use any index if we make the convention.

As for the final bullet point, i also used s because it is used by VF2 (it is the intermediate state of the algorithm, containing the mapping). But sure, we could just use the mapping m without s representing the intermediate state of the algorithm. 

As for the first bullet point, i will have to take a closer look at what you said so i completely understand your suggestion, because i didn't quite understand the purpose of the queue.

**Thanks once again!**

**Reply by kpetridis24** [reply]
I think that the functionality in lines 5-9 was not clear (i could be wrong though), so let me explain it better.

We take every node u of G1 (from the ordered list), in line 5
line 6 returns the list of candidates from G2, for node u.
line 7 checks if the matching can be established between (u,v) where v from the candidates (this check is performed for each candidate)
line 8 adds v to the mapping (M[u]=v) because we have a successful matching.
line 9 breaks the candidate search for u (as we just found a successful candidate), and takes the next u from G1, repeating the process

If a successful matching is found for all nodes u, the algorithm returns true. 
The only thing that is not quite clear yet is the case when, for some u, we run out of candidates and still no matching occured. 

You refer to removing v from M. Here is what i get from this:
If no candidate from G2 can be matched to the current u from G1, we should remove u from M. For example let's say the mapping goes like M = [1, 2, 3], so we have found a matching for nodes 1,2,3 of G1. Then we start searching for a matching for node 4 of G1 (let's assume it's the next one in the ordering). From all the candidates from G2, not a single one can be matched to node 4. So the mapping has to be developed following a different branch (and not 1->2->3->4), so we try another node from M, (for example 1->2->3->5 and try to match 5 with a node from G2 etc). 

**Do i get it right??**

**Reply by kpetridis24** [reply]
In more simple terms, we start with a full list containing the ordered nodes. We examine if a matching can be established for a node in the list (by taking its candidate from G2 each time). When we find a matching for some u from the list, we remove it from the list (add it to M) and start examining the nodes in order once again (starting from the beginning each time). For example,

**iter=1**
orderedNodes = (1, 2, 3, 4, 5)
nv = candidates for 1 -> no matching
nv = candidates for 2 -> matched with 7 (from G2) so M[2]=7

**iter=2**
orderedNodes = (1, 3, 4, 5)
nv = candidates for 1 -> no matching
nv = candidates for 3 -> matched to 8 (M[3]=8)

**iter=3**
orderedNodes = (1, 4, 5)
nv = candidates for 1 -> matched to 10 (M[1]=10)

**iter=4**
orderedNodes = (4, 5)
.
.
.
Until no nodes are left in the ordering. I believe that's what u meant when you suggested the queue. After i tested VF2, i realized that this is exactly how it works (but with a random ordering), so it should be the same thing but now using a specific order as computed by VF2++.

**Reply by dschult** [reply]
Another good algorithm that is very much like this is the [depth_first_search](https://networkx.org/documentation/stable/_modules/networkx/algorithms/traversal/depth_first_search.html#dfs_edges) set of functions.  The big idea is that you don't have to start from the beginning each time you back up. If you store an iterator of the adjacency on the queue, then you can just go to the next candidate without retracing your steps. 

If you think of the VF2 algorithm as searching a large tree of possibilities, we are doing a depth first search of the branches. We try to go all the way to a leaf... And if something doesn't work, we go back up and try the next way to get to a leave.  The bonus of VF2++ over VF2 is that we are choosing the hardest branches to make work first, so presumably we will run into impossible conditions early in the branch and be able to eliminate large portions of the tree. That may be just a naive version of the advantages, but that's my intuition about why the ordering works better.

Also, my understanding is that for **iter=1** if the first node doesn't have any candidates, we know the graphs are not isomorphic and we stop. If there are e.g. 4 candidates, we try the first one. Then we go to **iter=2**, and get the candidates for node 2 (with node 1 matched as already chosen). If there aren't any candidates left for node 2, we know the matching selected so far doesn't work, but this time we don't know the graphs are not isomorphic, because we still have more candidates to try for node 1. So, we go back up to **iter=1** and change the mapping so that 1 is mapped to the 2nd candidate in the 4 listed for node 1.  Heading back to **iter=2** this time we find 5 candidates for node 2. We select the first and add it to the map.  Then on to **iter=3** where we find 7 candidates and select one. Then **iter=4** find 3 candidates for node 4 and we select one. Then to **iter=5** where suppose there are no candidates. That means we go back to **iter=4** and try the next candidate there. This time, **iter=5** has only one candidate, and we select it. Moving to **iter=6** suppose there are no candidates. Then this branch can't work. So we go back to **iter=5**, but there are no candidates remaining there for node 5, so we go back up to **iter=4** and map node 4 to the next candidate on the iterator of candidates at that level. Then back down to the **iter=5** level. If we ever get to the **iter=len(G)** and find a candidate, we have found an isomorphism! If we ever go all the way back up to **iter=1** and do not have any candidates left, then we have shown that there is no isomorphism.   Basically, we're doing a depth-first-search through a virtual graph of possible mappings. At each level we add a node to the mapping and proceed to the next level to find what is possible with what we've locked-in so far. 

The queue stores information about each level of iteration.  Probably the node mapped at that level, the iterator of remaining candidates to try, and maybe some other information to help save computation... You'll have to see what is needed.  In the depth-first-search code, the queue stores the node at that level, the number of the level (called depth, but you could call it height if you like your trees to go up) and the unused iterator ready to provide the next candidate.

I don't know if this story helps, but it's how I think about it.

**Reply by kpetridis24** [reply]
This have been very helpful. Thank you very much. As it seems, we are indeed going to need a state object, storing the information at an intermediate level of the tree (node, remaining candidates) and we store every state in a queue, so when we go up the tree we just pop a previous state from the queue.

**Reply by kpetridis24** [reply]
Let State be an intermediate state of the process, storing the below information:
```
Class State:
  order             // The remaining nodes from G1 to be examined in the current state, as computed by the VF2++ ordering.
  u                   // The node from the order, currently being examined.
  M                 // The mapping, as extended until this state.
  candidates   // The remaining candidates to examine, for node u of the current state.
```

and 'states' be a list, storing State objects, i.e. the state of the algorithm in every level.

```
1.   input G1,G2
2.   states = [ ]
3.   s = State()
4.   s.order = orderVF2PP()     // the ordered nodes for the first state of the algorithm
5.   s.M = { }                            
6.   s.u = order[0]                    // we start examining the first node from the ordering
7.   while s.order not empty:
8.     u_candidates = find_candidates(s.u) 
9.     for v in u_candidates:  
10.     if Feasibility(s.u, v): 
11.       s.M[s.u] = v                                    // update mapping of current state
12.       s.candidates = u_candidates \ v    // remove the matched candidate, from the remaining candidates
13.       states.add(s)                                  // add the feasible state, to the states queue
14.       s_next = State()                             // create next state to continue the algorithm
15.       s_next.order = s.order \ s.u           // the already matched u, is excluded from the ordering 
16.       s_next.u = s_next.order[0]             // the new u to be examined is the next node of the ordering
17.       s_next.M = s.M
18.       s = s_next                           // next iteration of the algorithm is going to process the new state
19.       break                                 // terminate the candidate searching and repeat for new state
20.   if no feasible candidates found:
21.     s = states.pop()                           // the previous feasible state, now becomes the new state
22.     s.M.erase(s.u)        // remove the previously mapped pair, so a new one can be found from the remaining candidates
```
